### PR TITLE
bsunanda:Run2 alca35 Add IsoTrackFilter for normal physics events

### PR DESCRIPTION
**Pull request by bsunanda**
Use a new stream for AlCaReco

### CODE REVIEW

**Comment by mmusich** [comment]
please test

**Comment by bsunanda** [comment]
@mmusich May be I have to make it to your AlCaDB meeting on Monday for this requirement

File: Configuration/AlCa/python/autoAlca.py [hunk 1]
```diff
@@ -7,7 +7,8 @@
                   'DoubleMu'       : 'MuAlCalIsolatedMu+MuAlOverlaps+DtCalib+TkAlZMuMu',
                   'DoubleMuon'     : 'TkAlZMuMu+MuAlCalIsolatedMu+MuAlOverlaps+MuAlZMuMu+DtCalib',
                   'DoubleMuParked' : 'MuAlCalIsolatedMu+MuAlOverlaps+DtCalib+TkAlZMuMu',
-                  'HLTPhysics'     : 'SiStripCalMinBias+TkAlMinBias',
+                  'HLTPhysics'     : 'SiStripCalMinBias+TkAlMinBias+HcalCalIsoTrkFilter',
+                  'Cosmics'        : 'TkAlCosmics0T',
```
Comment (mmusich) [inline Configuration/AlCa/python/autoAlca.py]: @bsunanda the `Cosmics` key appears here twice. Please remove the unnecessary one.

**Comment by mmusich** [comment]
@bsunanda we can certainly discuss the details in the next AlCa/DB meeting. Is this a new AlCaReco that you want to add? As it is now, some things are missing.

**Comment by mmusich** [comment]
@bsunanda mainly you should add the new AlCaReco here : `Configuration/StandardSequences/python/AlCaRecoStreams_cff.py`

File: Calibration/HcalAlCaRecoProducers/python/ALCARECOHcalCalIsoTrkFilter_Output_cff.py [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+OutALCARECOHcalCalIsoTrkFilter_noDrop = cms.PSet(
+    # use this in case of filter available
+    SelectEvents = cms.untracked.PSet(
+        SelectEvents = cms.vstring('pathALCARECOHcalCalIsoTrk')
+    ),
+    outputCommands = cms.untracked.vstring( 
+        'keep *_hbhereco_*_*',
+        'keep *_reducedHcalRecHits_*_*',
+        'keep *_ecalRecHit_*_*',
+	'keep *_offlineBeamSpot_*_*',
```
Comment (mmusich) [inline Calibration/HcalAlCaRecoProducers/python/ALCARECOHcalCalIsoTrkFilter_Output_cff.py]: is it possible to get better indentation here?

File: Calibration/HcalAlCaRecoProducers/python/alcaIsoTracksFilter_cfi.py [hunk 3]
```diff
@@ -9,11 +9,11 @@
                                    HBHERecHitLabel   = cms.InputTag("hbhereco"),
                                    TriggerEventLabel = cms.InputTag("hltTriggerSummaryAOD","","HLT"),
                                    TriggerResultLabel= cms.InputTag("TriggerResults","","HLT"),
-                                   Triggers          = cms.vstring("HLT_PFJet40","HLT_PFJet80","HLT_PFJet140","HLT_PFJet200","HLT_PFJet260","HLT_PFJet320","HLT_PFJet320","HLT_PFJet400"),
+                                   Triggers          = cms.vstring(),
```
Comment (mmusich) [inline Calibration/HcalAlCaRecoProducers/python/alcaIsoTracksFilter_cfi.py]: Is the purpose of this to really accept any trigger that populates the PD? Do you foresee to run the new stream on HLTPhysics as well during 2016?

Comment (bsunanda) [inline Calibration/HcalAlCaRecoProducers/python/alcaIsoTracksFilter_cfi.py]: @mmusich Yes at this stage we do not care about the HLT

**Comment by mmusich** [comment]
please test

**Comment by mmusich** [comment]
@bsunanda sorry, you need also to update 
`Configuration/EventContent/python/AlCaRecoOutput_cff.py `

File: Calibration/HcalAlCaRecoProducers/python/ALCARECOHcalCalIsoTrkFilter_cff.py [hunk 4]
```diff
@@ -0,0 +1,22 @@
+import FWCore.ParameterSet.Config as cms
+
+#------------------------------------------------
+#AlCaReco filtering for HCAL isotrk:
+#------------------------------------------------
+
+from Calibration.HcalAlCaRecoProducers.alcaIsoTracksFilter_cfi import *
+
+import Alignment.CommonAlignmentProducer.AlignmentTrackSelector_cfi
+TkAlIsoProdFilter = Alignment.CommonAlignmentProducer.AlignmentTrackSelector_cfi.AlignmentTrackSelector.clone()
```
Comment (mmusich) [inline Calibration/HcalAlCaRecoProducers/python/ALCARECOHcalCalIsoTrkFilter_cff.py]: @bsunanda should't this be called TkAlIsoProd, to be consistent with the following lines?
I get this error when trying to run your PR merged on top of the last IB:
```
 File "/tmp/musich/CMSSW_8_0_X_2016-02-10-2300/python/Calibration/HcalAlCaRecoProducers/ALCARECOHcalCalIsoTrkFilter_cff.py", line 11, in <module>
    TkAlIsoProd.filter = False
NameError: name 'TkAlIsoProd' is not defined
```

**Comment by bsunanda** [comment]
@cmsbuild Please test

**Comment by mmusich** [comment]
@bsunanda do you plan to run this calibration also in Heavy Ions? I am not sure how many isolated tracks you can get there.

**Comment by bsunanda** [comment]
i don't think i will at this moment. i just saw the other isotrack was there - so i added that. have to re-evaluate by examining resources and benefits
________________________________

**Comment by bsunanda** [comment]
@mmusich Shall I remove this from HI?
________________________________
From: Sunanda Banerjee
Sent: 12 February 2016 10:17
To: cms-sw/cmssw; cms-sw/cmssw
Subject: RE: [cmssw] bsunanda:Run2 alca35 Add IsoTrackFilter for normal physics events (#13251)

i don't think i will at this moment. i just saw the other isotrack was there - so i added that. have to re-evaluate by examining resources and benefits
________________________________

**Comment by mmusich** [comment]
@bsunanda, no, it doesn't harm. It won't be run by default as you know, since the Tier-0 alca matrix gets configured in a different way.

File: Calibration/HcalAlCaRecoProducers/plugins/AlCaIsoTracksFilter.cc [hunk 5]
```diff
@@ -198,141 +198,152 @@ bool AlCaIsoTracksFilter::filter(edm::Event& iEvent, edm::EventSetup const& iSet
 			   << iEvent.bunchCrossing();
 
   //Step1: Find if the event passes one of the chosen triggers
-  trigger::TriggerEvent triggerEvent;
-  edm::Handle<trigger::TriggerEvent> triggerEventHandle;
-  iEvent.getByToken(tok_trigEvt_, triggerEventHandle);
-  if (!triggerEventHandle.isValid()) {
-    edm::LogWarning("HcalIsoTrack") << "Error! Can't get the product "
-				    << triggerEvent_.label() ;
+  bool ok(false);
```
Comment (mmusich) [inline Calibration/HcalAlCaRecoProducers/plugins/AlCaIsoTracksFilter.cc]: @bsunanda for the sake of code readibilty would it be possible to find a variable name more meaningful than "ok"?

Comment (bsunanda) [inline Calibration/HcalAlCaRecoProducers/plugins/AlCaIsoTracksFilter.cc]: @mmusich I shall name it differenly then
________________________________

**Comment by bsunanda** [comment]
@cmsbuild Please test

**Comment by mmusich** [comment]
+1


This PR:
   * adds the new `HcalCalIsoTrkFilter` AlCaReco based on the `AlCaIsoTracksFilter` EDF
   * changes the code for the EDFilter to deal with empty trigger list
   * adds in the `autoAlCa` matrix the new alcareco on the `HLTPhysics`,`JetHT`,`DoubleElectron` and `DoubleEG` PDs
   * The new AlCaReco is tested via WF 1000.0 in the short matrix through the `allForPrompt` list.

Analyzing the ouptput on 100 MinBias events from 2011:

```
$ edmEventSize -v -a 1000.0_RunMinBias2011A+RunMinBias2011A+TIER0+SKIMD+HARVESTDfst2+ALCASPLIT/HcalCalIsoTrkFilter.root

File 1000.0_RunMinBias2011A+RunMinBias2011A+TIER0+SKIMD+HARVESTDfst2+ALCASPLIT/HcalCalIsoTrkFilter.root Events 1
Branch Name | Average Uncompressed Size (Bytes/Event) | Average Compressed Size (Bytes/Event) 
recoTrackExtras_generalTracks__RECO. 63967 32240
recoTrackExtras_TkAlIsoProdFilter__RECO. 55579 29412
EcalRecHitsSorted_ecalRecHit_EcalRecHitsEB_RECO. 44704 23665
recoTracks_TkAlIsoProdFilter__RECO. 49535 19769
recoTracks_generalTracks__RECO. 49479 19713
HBHERecHitsSorted_hbhereco__RECO. 41412 19524
TrackingRecHitsOwned_generalTracks__RECO. 119244 18255
SiStripClusteredmNewDetSetVector_TkAlIsoProdFilter__RECO. 32532 18091
TrackingRecHitsOwned_TkAlIsoProdFilter__RECO. 119300 17592
SiPixelClusteredmNewDetSetVector_TkAlIsoProdFilter__RECO. 17232 10553
EcalRecHitsSorted_ecalRecHit_EcalRecHitsEE_RECO. 9786 6462
HBHERecHitsSorted_reducedHcalRecHits_hbhereco_RECO. 3780 3369
triggerTriggerEvent_hltTriggerSummaryAOD__HLT. 5243 3359
recoVertexs_offlinePrimaryVertices__RECO. 6364 3176
floats_generalTracks_MVAValues_RECO. 2880 2868
recoVertexsedmAssociation_offlinePrimaryVertices__RECO. 3164 2784
HFRecHitsSorted_reducedHcalRecHits_hfreco_RECO. 2658 2658
HORecHitsSorted_reducedHcalRecHits_horeco_RECO. 2658 2658
edmTriggerResults_GammaJetProd_TriggerResults::HLT_RECO. 2643 2641
intedmValueMap_offlinePrimaryVertices__RECO. 2908 2510
floatedmValueMap_offlinePrimaryVertices__RECO. 2508 2508
edmTriggerResults_TriggerResults__HLT. 3019 2465
edmTriggerResults_TriggerResults__RECO. 2479 2442
recoBeamSpot_offlineBeamSpot__RECO. 2590 2442
uchars_generalTracks_QualityMasks_RECO. 2507 2415
EventProductProvenance 3373 2119
EventAuxiliary 738 220
EventSelections 700 182
BranchListIndexes 638 112
```

@bsunanda is expect to give details of the event rates and output product sizes at the next AlCa/DB meeting.

File: Calibration/HcalAlCaRecoProducers/python/ALCARECOHcalCalIsoTrkFilter_Output_cff.py [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+    ),
+    outputCommands = cms.untracked.vstring( 
+        'keep *_hbhereco_*_*',
+        'keep *_reducedHcalRecHits_*_*',
+        'keep *_ecalRecHit_*_*',
+        'keep *_offlineBeamSpot_*_*',
+        'keep *_hltTriggerSummaryAOD_*_HLT',
+        'keep *_TriggerResults_*_*',
+        'keep edmTriggerResults_*_*_*',
+        'keep triggerTriggerEvent_*_*_*',
```
Comment (davidlange6) [inline Calibration/HcalAlCaRecoProducers/python/ALCARECOHcalCalIsoTrkFilter_Output_cff.py]: @bsunanda - please add the module names you desire to keep for all products.

**Comment by bsunanda** [comment]
@cmsbuild Please test

**Comment by bsunanda** [comment]
@cmsbuild Please test

**Comment by bsunanda** [comment]
@mmusich Could you approve this so that we an request the filter to be deployed for 2016 data

**Comment by mmusich** [comment]
+1

**Pull request merged**